### PR TITLE
Added some ARI commands and events

### PR DESCRIPTION
**Pull request by znoble360**
I added events for recording_started and recording_finished, 
I added the snoop and get recording ari commands, 

I updated readme accordingly,
and I updated the simple_client example to include the new events I added.

### CODE REVIEW

**Comment by adambezecny** [comment]
hi Zachary,

thank you for your contribution. Can you please instead of extending simple_client example create dedicated example with new features introduced by your PR?  It would be fine if it demonstrates some reproducible use case you can do with these new features. It will help me to test it. thanks. 

regards,

Adam

**Comment by znoble360** [comment]
I can see about doing that

File: Cargo.toml [hunk 1]
```diff
@@ -27,6 +27,10 @@ log = "0.4.20"
 chrono = "0.4.31"
 rand = "0.8.5"
 lazy_static = "1.4.0"
+percent-encoding = "2.3.1"
 
 [dev-dependencies]
 env_logger = "0.10.1"
+
+[features]
+parse-event-datetimes = []
```
Comment (adambezecny) [inline Cargo.toml:36]: what exactly is this features used for? provide some clarification please.

Comment (adambezecny) [reply Cargo.toml:36]: ok found it, got it :)

File: src/models/channels.rs [hunk 2]
```diff
@@ -203,3 +205,23 @@ pub struct RtPstat {
...
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, Default)]
+pub enum Direction {
+    #[default]
+    None,
+    Both,
+    Out,
+    In
+}
+impl fmt::Display for Direction {
```
Comment (adambezecny) [inline src/models/channels.rs:217]: can you add some comment describing this enum and its options?

**Comment by adambezecny** [comment]
hi Zachary,

first of all apologies for such a long delays, I am pretty busy these days and it is not easy to grasp some time for things like PR reviews on my personal project(s) :)

in general I am OK with this change. What I am missing is some dedicated example. I would like to avoid extending general example (**simple_client.rs**) and prefer smaller dedicated examples showing some use case/feature in action (e.g. **recording.rs**). I suggest moving events recording started/ended into recording.rs example where it will make more sense. 

Is it possible somehow to demonstrate snoop feature and get recording? Can you think of some use case we could do on local dockerized asterisk setup (see readme in examples)?

**Comment by adambezecny** [comment]
regarding snooping, if I understand it correctly it could be used for something like sending channel audio output into speech to text API and then spying to send back some synthetized response? Do you think you could prepare some simple dedicated example for this use case? I remember using external plugin (audio fork) to tap to channel and receive audio data. If snooping could do the same natively that would be great. looking forward hearing from you more.

**Comment by adambezecny** [comment]
https://docs.asterisk.org/Development/Reference-Information/Asterisk-Framework-and-API-Examples/External-Media-and-ARI/
https://docs.asterisk.org/Latest_API/API_Documentation/Dialplan_Applications/EAGI/
https://docs.asterisk.org/Latest_API/API_Documentation/Dialplan_Applications/JACK/

https://github.com/asterisk/asterisk-external-media

https://community.asterisk.org/t/how-to-snoop-on-a-channel-in-ari/100978/4

**Comment by znoble360** [comment]
Sorry, I haven't been paying a whole lot of attention to this for the time being, I also have been doing other things. The use case I was using snoop for was a very similar use case to mixmonitors where multiple channels can be condensed down to one for recording purposes. the reason I can't use mixmonitor for that is because I think it doesn't show up as a recording in the recording api, but snooping plays nicer with the api.

**Comment by adambezecny** [comment]
ok, so regarding new events **recording_started/recording_finished** please put them into dedicated example (**recordings.rs**) which is already there.  main example is too big and crowded, I do not want to put any new stuff here.

regarding **snoop**: I am still little bit int he mist. I am not really Asterisk expert and not aware of most of Asterisk advanced features/configurations/etc (including mixmonitor :-) ). If new API should be merged I would like really to have dedicated example where snoop will be demonstrated. It should work against this asterisk configuration: https://github.com/jabber-tools/asterisk-ari-client-rs/tree/main/examples/ast-docker

If you need to add some extension or other stuff here np. But I would like to have clear and easy to use demonstration of this new feature.

regards,

Adam

**Review by adambezecny — APPROVED** [review]
lgtm

**Comment by adambezecny** [comment]
hi Zachary,

thank you for your contribution, I am merging now. Give me some time to roll out new version as I need to do some minor changes. Once available on **crates.io** I will ping you.

Please note I am also maintaining following libraries

- https://crates.io/crates/cognitive-services-speech-sdk-rs
- https://crates.io/crates/google-cognitive-apis

Maybe you will find them useful for your use case(s) or maybe you even submit some PR :)


regards,

Adam

**Comment by adambezecny** [comment]
version **0.1.4** has been just released, details here https://github.com/jabber-tools/asterisk-ari-client-rs/releases/tag/0.1.4

once again, thanks for your contribution!

regards,

Adam

**Pull request merged**